csrc/mmdeploy/net/coreml/coreml_net.mm

```diff
@@ -0,0 +1,321 @@
+  const std::vector<mmdeploy::Tensor> *inputs_;
+}
+
+- (instancetype)initWithInputs:(const std::vector<mmdeploy::Tensor> &)inputs;
+- (NSInteger)count;
+- (id<MLFeatureProvider>)featuresAtIndex:(NSInteger)index;
+@end
+
+@implementation MMBatchTensorFeatureProvider
+
+- (instancetype)initWithInputs:(const std::vector<mmdeploy::Tensor> &)inputs {
+  inputs_ = &inputs;
+  return self;
+}
+
+- (NSInteger)count {
+  return (*inputs_)[0].shape(0);
+}
+
+- (id<MLFeatureProvider>)featuresAtIndex:(NSInteger)index {
+  MLDictionaryFeatureProvider *feature;
+  NSMutableDictionary<NSString *, id> *input_dict =
+      [[NSMutableDictionary<NSString *, id> alloc] init];
+
+  for (auto x : *inputs_) {
+    auto in = x.Slice(index);
+    NSMutableArray *shape = [[NSMutableArray alloc] init];
+    for (const auto dim : in.shape()) {
+      [shape addObject:[NSNumber numberWithLongLong:dim]];
+    }
+
+    NSMutableArray *strides = [[NSMutableArray alloc] init];
+    int64_t stride = 1;
+    for (int i = in.shape().size() - 1; i >= 0; i--) {
+      [strides insertObject:[NSNumber numberWithLongLong:stride] atIndex:0];
+      stride *= in.shape()[i];
+    }
+
+    MLMultiArrayDataType data_type = MLMultiArrayDataTypeFloat32;
+    NSError *error = nil;
+    MLMultiArray *mlArray =
+        [[MLMultiArray alloc] initWithDataPointer:in.data()
+                                            shape:shape
+                                         dataType:data_type
+                                          strides:strides
+                                      deallocator:(^(void *){
+                                                  })error:&error];
+    if (error != nil) {
+      MMDEPLOY_ERROR("init MLMultiArray failed with key: {}, error message: {}",
+                     in.name(), [[error localizedDescription] UTF8String]);
+    }
+
+    NSString *key = [NSString stringWithUTF8String:in.name()];
+    input_dict[key] = mlArray;
+  }
+
+  NSError *error = nil;
+  feature = [[MLDictionaryFeatureProvider alloc] initWithDictionary:input_dict
+                                                              error:&error];
+  if (error != nil) {
+    MMDEPLOY_ERROR("init MLDictionaryFeatureProvider failed with index: {}, "
+                   "error message: {}",
+                   index, [[error localizedDescription] UTF8String]);
+  }
+
+  return feature;
+}
+
+@end
+
+namespace mmdeploy {
+namespace coreml {
+
+static Result<void> CheckInputOutputFeatureType(MLFeatureType type) {
+  if (type != MLFeatureTypeMultiArray) {
+    MMDEPLOY_ERROR("unsupported feature type: {}", type);
+    return Status(eInvalidArgument);
+  }
+  return success();
+}
+
+static TensorShape to_shape(NSArray<NSNumber *> *shape) {
+  TensorShape _shape;
+  for (int i = 0; i < shape.count; i++) {
+    _shape.push_back(shape[i].intValue);
+  }
+  return _shape;
+}
+
+static Result<DataType> ConvertElementType(MLMultiArrayDataType type) {
+  switch (type) {
+  case MLMultiArrayDataTypeFloat32:
```

Comment:
Is it possible that `type` can be `int8`?

Reply:
It can be float64, . We only think the common types for now.